README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# elastes

<!-- badges: start -->

[![R-CMD-check](https://github.com/mpff/elastes/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/mpff/elastes/actions/workflows/R-CMD-check.yaml)
[![codecov](https://codecov.io/gh/mpff/elastes/branch/main/graph/badge.svg?token=p0xOHfDpnk)](https://app.codecov.io/gh/mpff/elastes)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

elastes is a R package that provides functions for the computation of
functional elastic shape means over sets of open planar curves. The
package is particularly suitable for settings where these curves are
only sparsely and irregularly observed. It uses a novel approach for
elastic shape mean estimation, where planar curves are treated as
complex functions and a full Procrustes mean is estimated from the
corresponding smoothed Hermitian covariance surface. This is combined
with the methods for elastic mean estimation proposed in [Steyer,
Stöcker, Greven (2022)](https://doi.org/10.1111/biom.13706). See
[Stöcker et. al. (2022)](https://doi.org/10.48550/arXiv.2203.10522) for
details.

## Installation

The simplest way to obtain elastes is to install it directly from CRAN.
Type the following command in your R console:

``` r
install.packages("elastes")
```

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("mpff/elastes")
```

## Example

Calculate a smooth elastic shape mean for sparse spirals.

``` r
library(elastes)

# define spiral curve
curve <- function(t){
  rbind(t*cos(13*t), t*sin(13*t))
}

# randomly draw sparse spirals with noise
set.seed(18)
data_curves <- lapply(1:10, function(i){
  m <- sample(10:15, 1)
  delta <- abs(rnorm(m, mean = 1, sd = 0.05))
  t <- cumsum(delta)/sum(delta)
  data.frame(t(curve(t)) + 0.07*t*matrix(cumsum(rnorm(2*length(delta))),
             ncol = 2))
})

# apply random rotation and scaling to spiral curves
rand_scale <- function(curve){ ( 0.5 + runif(1) ) * curve }
rand_rotate <- function(curve){
  names <- colnames(curve)
  theta <- 2*pi*runif(1)
  mat <- matrix(c(cos(theta), sin(theta), -sin(theta), cos(theta)), nrow = 2, ncol = 2)
  curve.rot <- as.matrix(curve) %*% t(mat)
  curve.rot <- as.data.frame(curve.rot)
  colnames(curve.rot) <- names
  return(curve.rot)
}
data_curves <- lapply(data_curves, rand_scale)
data_curves <- lapply(data_curves, rand_rotate)

# compute smooth procrustes mean using a 2nd order penalty in the estimation
knots <- seq(0,1, length = 11)
elastic_shape_mean <- compute_elastic_shape_mean(
  data_curves,
  knots = knots,
  type = "smooth",
  penalty = 2
)
plot(elastic_shape_mean)
```

![](man/figures/README-unnamed-chunk-4-1.png)<!-- -->


NEWS.md
# elastes 0.1.7

* Updated e-mail and copyright holder in `DESCRIPTION` file.
* Add appropriate PKGNAME-package \alias as per "Documenting packages" in R-exts.


# elastes 0.1.6

* Changed T/F to TRUE/FALSE in function's default parameters.
* Documented missing return values in exported functions.
* Removed 'par()' call in 'elastes.Rmd' vignette.

# elastes 0.1.5

* Added a `NEWS.md` file to track changes to the package.
* Fixed a bug where gram matrix was wrongly calculated with the piecewise constant basis and irregular knots (#10).
* Updated DESCRIPTION and documentation.
* Added tests for plotting and evaluating means. Clean up existing tests by removing duplicate and unnecessary computations.
* Added a vignette for getting started.
* Created a pkgdown site.


tests/testthat.R
library(testthat)
library(elastes)

test_check("elastes")


tests/testthat/test-basis.R
knots <- c(0, 1)

test_that("Test can get inner and outer knots for both types", {
  expect_equal(get_knots(knots, "polygon"), c(0, 1))
  expect_equal(get_knots(knots, "smooth"), c(-1, 0, 1, 2))
})

test_that("Test can get design matrix for both types", {
  t <- c(0, 1)
  dp <- matrix(c(1, 1), nrow=2)
  expect_equal(make_design(t, knots, "polygon"), dp)
  ds <- matrix(c(1, 0, 0, 1), nrow=2)
  expect_equal(make_design(t, knots, "smooth"), ds)
})

test_that("Test can get Gram matrix for equidistant knots", {
  gp <- matrix(1)
  expect_equal(get_gram_matrix(knots, "polygon"), gp)
  gs <- matrix(c(1/3, 1/6, 1/6, 1/3), nrow=2)
  expect_equal(get_gram_matrix(knots, "smooth"), gs)
})

test_that("Test can get Gram matrix for type 'polygon' with irregular knots", {
  knots_irreg <- c(0, 0.1, 1)
  gip <- matrix(c(0.1, 0, 0, 0.9), nrow=2)
  expect_equal(get_gram_matrix(knots_irreg, "polygon"), gip)
})


tests/testthat/test-utils.R
data_curve <- data.frame(x1 = c(0,0.5,1), x2 = c(0,0,0))

test_that("Test can calculate polygon length", {
  expect_equal(get_polygon_length(data_curve), 1)
})

test_that("Test can calculate curve center", {
  expect_equal(get_center(data_curve),  c(x1 = 0.5, x2 = 0))
})

test_that("Test distance between same curves is zero",{
  srv_data_curve <- elasdics::get_srv_from_points(data_curve)
  knots <- c(0,0.5,1)
  type <- "polygon"
  coefs <- as.matrix(srv_data_curve[,-1])
  distance <- get_distance(
    srv_curve = function(t) t(make_design(t, knots=knots, type=type) %*% coefs),
    s = c(srv_data_curve$t, 1),
    q = t(srv_data_curve[,-1])
  )
  expect_equal(distance, 0, tolerance=1e-6)
})


R/utils.R
#' Calculate the polygon length of a curve
#'
#' @param curve a \code{data.frame} with observed points in each row. Each
#'   variable is one coordinate direction. If there is a variable \code{t},
#'   \code{t_optim} or \code{id}, it is treated as the time parametrization, not
#'   as an additional coordinate.
#' @return The length of \code{curve}, treating it as a polygon.

get_polygon_length <- function(curve) {
  coord_idx <- !(colnames(curve) %in% c("t", "t_optim", "id"))
  curve <- curve[,coord_idx]
  dx <- diff(curve[,1])
  dy <- diff(curve[,2])
  sum(sqrt(dx^2 + dy^2))
}


#' Calculate the center of a curve
#'
#' @param curve a \code{data.frame} with observed points in each row. Each
#'   variable is one coordinate direction. If there is a variable \code{t},
#'   \code{t_optim} or \code{id}, it is treated as the time parametrization, not
#'   as an additional coordinate.
#' @return The average of observed points in \code{curve}.

get_center <- function(curve) {
  coord_idx <- !(colnames(curve) %in% c("t", "t_optim", "id"))
  curve <- curve[,coord_idx]
  colMeans(curve)
}


#' @title Distance to a smooth curve
#' @description Finds the distance of a discrete open srv curve to a smooth curve
#' @param srv_curve srv transformation of the smooth curve, needs to be vectorized
#' @param s time points for q, first has to be 0, last has to be 1
#' @param q square root velocity vectors, one less than time points in s
#' @param eps convergence tolerance
#' @return distance between srv_curve and q

get_distance <- function(srv_curve, s, q, eps = 10*.Machine$double.eps){
  p_integrand <- function(t){sapply(t, function(t) sum(srv_curve(t)^2))}
  p_norm <- stats::integrate(p_integrand, 0, 1, stop.on.error =FALSE)$value  # should be 1
  q_norm <- sum(q^2 %*% diff(s))  # should be 1
  pq_prod <- sapply(1:(length(s)-1), function(i) {
    stats::integrate(function(t) t(q[,i]) %*% srv_curve(t), s[i], s[i+1], stop.on.error = FALSE)$value
  })
  dist2 <- p_norm + q_norm - 2 * sum(pq_prod)
  if(dist2 < 0) dist2 <- 0
  sqrt(dist2)
}


R/elastes.R
#' elastes: Elastic Full Procrustes Means for Sparse and Irregular Planar Curves
#'
#' Provides functions for the computation of functional elastic shape
#' means over sets of open planar curves. The package is particularly suitable for
#' settings where these curves are only sparsely and irregularly observed. It uses
#' a novel approach for elastic shape mean estimation, where planar curves are
#' treated as complex functions and a full Procrustes mean is estimated from the
#' corresponding smoothed hermitian covariance surface, which is combined with the
#' methods for elastic mean estimation proposed in Steyer, Stöcker, Greven	(2022).
#' See Stöcker et. al. (2022) for details on the method.
#'
#' Compute a mean for a set of observed curves: \code{\link{compute_elastic_shape_mean}}
#'
#' @docType package
#' @name elastes
NULL


R/fit_alignment_proc2d.R
#' @title Optimal rotation and scaling alignment to a smooth curve
#' @description Finds optimal rotation and scaling alignment for a discrete open srv curve to a smooth curve
#' @param q complex srv curve with parametrization, needs to be vectorized.
#' The result of a call to \code{get_model_data_complex}
#' @param type spline degree
#' @param knots basis knots
#' @param var_type either "smooth" or "constant" measurement error in cov_fit object
#' @param coefs.compl complex coefficients of smooth curve
#' @param method temp
#' @param cov_fit temp
#' @param pca temp
#' @param L temp
#' @return optimal rotation G and scaling b

fit_alignment_proc2d <- function(q, type, knots, var_type, coefs.compl, method, cov_fit, pca, L){

  if (method == "smooth") {

    # Get covariance eigen-basis vectors and values.
    V <- pca$vectors
    Lmbd <- pca$values

    # Restrict basis to ensure positive definiteness (see issue #2)
    Lmbd <- Lmbd[pca$values > 0]
    V <- V[ , pca$values >0]

    # Ensure correct Lmbd.inv matrix when only one eigenvalue (see issue #8)
    Lmbd.inv <- if(length(Lmbd) == 1) matrix(1/Lmbd) else diag(1/Lmbd)

    # Construct the measurement error variance
    if(var_type == "zero"){
      T_ <- rep(0, length(q$m_long))
    } else {
      s_index <- ifelse(var_type == "smooth", 2, 1)
      T_ <- stats::predict(cov_fit$re, data.frame(t=q$m_long, s=q$m_long, st=1), type = "terms")[,s_index]
    }

    if(all(T_ > 0)){

      # Construct design matrix in eigen-basis. (Note: L <- chol(G.inv) for Gram-Matrix G of the mean basis)
      q_B <- make_design(q$m_long, knots, type = type)
      E <- q_B %*% t(L) %*% V

      # Construct inverse error variance matrix.
      T.inv <- diag(1/T_)

      # Conditional score covariance.
      S <- solve(t(Conj(E)) %*% T.inv %*% E + Lmbd.inv)

      # Estimate score vector.
      z <- S %*% t(Conj(E)) %*% T.inv %*% q$q_m_long

    } else {

      # Get idxs of pos./neg. part of error variance
      ip <- which(T_ > 0)
      i0 <- which(T_ <= 0)

      if(length(ip) > 0){
        # Construct design matrix in eigen-basis. (Note: L <- chol(G.inv) for Gram-Matrix G of the mean basis)
        q_Bp <- make_design(q$m_long[ip], knots, type = type)
        q_B0 <- make_design(q$m_long[i0], knots, type = type)
        Ep <- q_Bp %*% t(L) %*% V
        E0 <- q_B0 %*% t(L) %*% V

        # Construct inverse error variance matrix (of pos. part).
        Tp.inv <- diag(1/T_[ip])

        # Construct QR-decomp of E0
        QR <- qr(t(E0))
        Q <- qr.Q(QR, complete = T)
        M <- Q[,1:QR$rank]

        # Estimated deterministic part of score vector.
        z0 <- M %*% solve( t(Conj(M)) %*% t(Conj(E0)) %*% E0 %*% M) %*% t(Conj(M)) %*% t(Conj(E0)) %*% q$m_long[i0]

        if(QR$rank < ncol(Q)){
          N <- Q[,(QR$rank + 1):ncol(Q)]

          # Conditional score covariance.
          S <- solve( t(Conj(N)) %*% (t(Conj(Ep)) %*% Tp.inv %*% Ep + Lmbd.inv) %*% N)

          # Estimated random part of score vector.
          zp <- N %*% S %*% t(Conj(N)) %*% ( t(Conj(Ep)) %*% Tp.inv %*% (q$m_long[ip] - Ep %*% z0) - Lmbd.inv %*% z0)
        } else {
          N <- 0; S <- 0; zp <- 0
        }

      } else {
        q_B0 <- make_design(q$m_long[i0], knots, type = type)
        E0 <- q_B0 %*% t(L) %*% V
        QR <- qr(t(E0))
        Q <- qr.Q(QR, complete = T)
        M <- Q[,1:QR$rank]
        z0 <- M %*% solve( t(Conj(M)) %*% t(Conj(E0)) %*% E0 %*% M) %*% t(Conj(M)) %*% t(Conj(E0)) %*% q$m_long[i0]
        N <- 0; S <- 0; zp <- 0
      }

      # Estimate score vector.
      z <- zp + z0

    }

    # Prepare output
    w <- Conj(z[1])
    l <- Re(c(sum(diag(S)) + t(Conj(z)) %*% z))  # length of original curve
    q_coefs <- t(L) %*% V %*% z

  } else {

    m <- get_intervals(q$m_long)

    mean_func <- function(t) {t(make_design(t, knots = knots, type = type) %*% coefs.compl)}
    qm_ints <- sapply(1:(length(m)-1), function(i) {
      re <- stats::integrate(function(s) Re(Conj(q$q_m_long[i]) * mean_func(s)), m[i], m[i+1], rel.tol = 0.01)$value
      im <- stats::integrate(function(s) Im(Conj(q$q_m_long[i]) * mean_func(s)), m[i], m[i+1], rel.tol = 0.01)$value
      re + 1i * im
    })

    w <- sum(qm_ints)
    l <- sum(diff(m) * Conj(q$q_m_long) * q$q_m_long)
    q_coefs <- NULL
    S <- NULL

  }

  # Check length is real and positive. (See issue #2)
  if(Mod(l) != Re(l)){
    stop(paste0("In calculation of curve length: Length is negative or imaginary. l = ", l))
  }

  # Return
  list("w" = w, "l" = Re(l), "q_coefs" = q_coefs, "S" = S)
}


get_intervals <- function(m){
  out <- rep(0, length(m) + 1)
  for(i in 2:length(out)){
    out[i] <- out[i-1] + 2*(m[i-1] - out[i-1])
  }
  return(out)
}


R/fit_mean.R
#' Mean estimation for open planar curves.
#' @name fit_mean
#' @description Fits an elastic full Procrustes mean for open, planar curves.
#' Is usually called from \code{\link{compute_elastic_shape_mean}}.
#' @param srv_data_curves list of \code{data.frame}s with srv vectors in each row.curves
#' @param knots set of knots for the mean spline curve
#' @param type if "smooth" linear srv-splines are used which results in a differentiable mean curve
#' if "polygon" the mean will be piecewise linear.
#' @param penalty the penalty to use in the covariance smoothing step. use '-1' for no penalty.
#' @param pfit_method (experimental) "smooth" or "polygon"
#' @param var_type (experimental) assume "smooth", "constant" or "zero" measurement-error variance along t
#' @param smooth_warp (experimental) controls the weighting of original and smoothed observations
#' over the iterations, if pfit_method == "smooth".
#' @param max_iter maximal number of iterations
#' @param verbose print iterations
#' @param eps the algorithm stops if L2 norm of coefficients changes less
#' @param cluster a cluster object for use in the \code{bam} call
#' @return a \code{list} with entries
#'   \item{type}{"smooth" or "polygon"}
#'   \item{coefs}{\code{coefs} srv spline coefficients of the estimated mean}
#'   \item{knots}{spline knots}
#'   \item{penalty}{penalty used in the covariance estimation}
#'   \item{distances}{distances to mean}
#'   \item{fit}{a \code{list} containing
#'       \code{t_optims}{optimal parametrizations}
#'       \code{G_optims}{optimal rotations}
#'       \code{b_optims}{optimal scalings}
#'       \code{n_optims}{optimal re-normalization}
#'       \code{n_iter}{number of iterations until convergence}
#'       \code{gram} the mean basis Gram matrix,
#'       \code{cov_fit} the covariance smoothing objects in the final iteration,
#'       \code{cov_pca} cov coef matrix pca object in the final iteration and
#'       \code{pfit_coefs} the mean basis coefs of smoothed pfits in the final iteration}
#' @import sparseFLMM

fit_mean <- function(srv_data_curves, knots, penalty, var_type, pfit_method, max_iter, type, eps, cluster, verbose, smooth_warp){
  # Initial parametrisation, rotation, scaling and coefs.
  t_optims <- lapply(srv_data_curves, function(srv_data_curve){
    c(srv_data_curve$t, 1)
  })
  G_optims <- as.list(rep(0, length(srv_data_curves)))
  b_optims <- as.list(rep(1, length(srv_data_curves)))
  l_optims <- as.list(rep(1, length(srv_data_curves)))
  l_optims_old <- as.list(rep(1, length(srv_data_curves)))
  coefs <- 0
  pfit_coefs <- as.list(rep(0, length(srv_data_curves)))
  distances <- rep(1, length(srv_data_curves))

  # Get Gram and orthogonal trafo matrix for the mean basis.
  G <- get_gram_matrix(knots, type)
  G.inv <-solve(G)
  chol.G <- chol(G)
  chol.G.inv <- chol(G.inv)
  L <- chol(G.inv)
  L.inv <- solve(L)

  # Make Integration Grid
  h = 0.01
  arg.grid = seq(0, 1, by=h)

  #iterate procrustes mean fit and warping on procrustes fits
  for (i in 1:max_iter){

    if(verbose) message(paste0("  Iteration ", i, "..."))

    model_data_complex <- get_model_data_complex(t_optims, srv_data_curves, knots, type)

    # Check for inf in model_data_complex and drop.
    if(any(!is.finite(Re(model_data_complex$q_m_long)))) {
      drops <- model_data_complex[!is.finite(Re(model_data_complex$q_m_long)), ]
      warning(paste("    Warning: Dropping", nrow(drops), "point(s) in mean estimation."))
      model_data_complex <- model_data_complex[is.finite(Re(model_data_complex$q_m_long)), ]
    }

    # Build complex response on s,t-grid per curve
    cov_dat <- lapply(split(model_data_complex, model_data_complex$id), function(x) {
      co <- utils::combn(seq_len(nrow(x)), 2)
      di <- t(matrix(c(seq_len(nrow(x)), seq_len(nrow(x))), ncol=2))  # Include diagonal
      cb <- cbind(di, co)
      combs <- cb[,order(cb[1,], cb[2,])]
      data.frame(
        qq = Conj(x$q_m_long[combs[1,]]) * x$q_m_long[combs[2,]],
        s = x$m_long[combs[1,]],
        t = x$m_long[combs[2,]],
        st = as.numeric(x$m_long[combs[1,]] == x$m_long[combs[2,]])
      )
    })
    cov_dat <- do.call(rbind, cov_dat)

    # Tensor product p-spline smoothing of the complex covariance surface.
    cov_fit <- smooth_cov_surface(cov_dat, knots, type, penalty, var_type, cluster)

    # Get unconstrained basis coef matrix from cov fit.
    beta.mat <- get_complex_coef_matrix(cov_fit)
    beta.mat.inv <- solve(beta.mat)

    # Calculate mean from pca on coefficient matrix
    pca <- eigen(t(L.inv) %*% beta.mat %*% L.inv)
    coefs.compl <- t(L) %*% pca$vectors[,1]

    # Get norm of coefs.
    coefs.norm <- sqrt(t(Conj(coefs.compl)) %*% G %*% coefs.compl)
    coefs.compl <- coefs.compl/c(coefs.norm)

    # Save old coefs, build (x,y) coefs from complex.
    coefs_old <- coefs
    coefs <- as.matrix(data.frame(q_m_long.X1 = Re(coefs.compl), q_m_long.X2 = Im(coefs.compl)))

    # Calculate procrustes fits of the warped srv data curves
    srv_procrustes_curves <- lapply(1:length(srv_data_curves), function(j) {

      q <- model_data_complex[model_data_complex$id == j,]

      # Calculate Scalar products qq and qm (wird noch aufgeräumt!!)
      pfit_prods <- fit_alignment_proc2d(q, type, knots, var_type, coefs.compl, method = pfit_method, cov_fit, pca, L)

      # Get alignment results
      w <- pfit_prods$w  # Conj(z1)
      l <- pfit_prods$l  # length(beta)
      pfit_coefs[[j]] <<- pfit_prods$q_coefs

      # Save scaling and rotation alignment, save re-normalization.
      b_optims[[j]] <<- Mod(w)
      G_optims[[j]] <<- Arg(w)
      l_optims_old[[j]] <<- l_optims[[j]]  # b_optims and l_optims are from diff. iterations!
      l_optims[[j]] <<- l_optims[[j]] * l

      # Apply alignment.
      # SRV curve to complex.
      srv_complex <- complex(real = srv_data_curves[[j]][,2], imaginary = srv_data_curves[[j]][,3])

      # Update normalization of SRV curve.
      srv_complex <- c(1/sqrt(l)) * srv_complex
      srv_data_curves[[j]][,2] <<- Re(srv_complex)
      srv_data_curves[[j]][,3] <<- Im(srv_complex)

      # Blend between smoothed and original curves for warping alignment step.
      if(pfit_method == "smooth"){
        q_coefs_norm <- sqrt(t(Conj(pfit_prods$q_coefs)) %*% G %*% pfit_prods$q_coefs)
        q_coefs_normed <- pfit_prods$q_coefs / c(q_coefs_norm)
        srv_complex_smooth <- make_design(srv_data_curves[[j]][,1], knots, type) %*% q_coefs_normed
        srv_complex <- (1-smooth_warp(i)) * srv_complex + smooth_warp(i) * srv_complex_smooth
      }

      # Calculate rotation aligned and re-normalized SRV curve.
      pfit <- c(w/Mod(w)) * srv_complex  # rotation alignment

      # Return SRV Procrustes Fit.
      srv_procrustes_curve <- data.frame(t = srv_data_curves[[j]][,1], X1 = Re(pfit), X2 = Im(pfit))

      # Update distance to mean
      if(pfit_method == "smooth"){
        distances[[j]] <<- sqrt(1 - Mod(as.matrix(pfit_prods$S)[1,1] + c(w * Conj(w))/l))
      } else {
        distances[[j]] <- get_distance(
          srv_curve = function(t) t(make_design(t, knots=knots, type=type) %*% coefs),
          s = c(srv_procrustes_curve$t, 1),
          q = t(srv_procrustes_curve[,-1]),
          eps = eps*100/i
        )
      }

      srv_procrustes_curve
    })

    #Stop if coefficients didn't change by much.
    stop_crit <- sum((coefs - coefs_old)^2)/sum(coefs^2)
    if(stop_crit < eps | max_iter == 0){
      rownames(coefs) <- NULL
      colnames(coefs) <- colnames(srv_data_curves[[1]][,-1])
      fit_object <- list(
        "n_iter" = i,
        "gram" = G,
        "cov_fit" = cov_fit,
        "cov_coef" = beta.mat,
        "cov_pca" = pca,
        "pfit_coefs" = pfit_coefs,
        "G_optims" = G_optims,
        "b_optims" = b_optims,
        "l_optims" = l_optims_old
      )
      return(list("type" = type, "knots" = knots, "penalty" = penalty,
                  "var_type" = var_type, "pfit_method" = pfit_method, "smooth_warp" = smooth_warp,
                  "coefs" = coefs, "t_optims" = t_optims, "fit" = fit_object, "distances" = distances))
    }

    # align parametrization to mean
    t_optims <- get_optimal_t(srv_procrustes_curves, coefs, t_optims, type, knots, eps, i)

  }

  warning("Stopping criteria eps has not been reached! Consider more iterations max_iter")
  rownames(coefs) <- NULL
  colnames(coefs) <- colnames(srv_data_curves[[1]][,-1])
  fit_object <- list(
    "n_iter" = i,
    "gram" = G,
    "cov_fit" = cov_fit,
    "cov_coef" = beta.mat,
    "cov_pca" = pca,
    "pfit_coefs" = pfit_coefs,
    "G_optims" = G_optims,
    "b_optims" = b_optims,
    "l_optims" = l_optims_old
  )
  return(list("type" = type, "knots" = knots, "penalty" = penalty,
              "var_type" = var_type, "pfit_method" = pfit_method, "smooth_warp" = smooth_warp,
              "coefs" = coefs, "t_optims" = t_optims, "fit" = fit_object, "distances" = distances))
}


# Get complex srv data curves, with additional oversampling for identifiability.
get_model_data_complex <- function(t_optims, srv_data_curves, knots, type){

  if (type == "polygon") {
    q_m <- lapply(1:length(srv_data_curves), function(j) {
      curve <- data.frame(t = t_optims[[j]], elasdics::get_points_from_srv(srv_data_curves[[j]]))
      curve_at_knots <- cbind(t = knots, get_evals(curve, t_grid = knots))
      elasdics::get_srv_from_points(curve_at_knots)[, -1, drop = FALSE]
    })
    m <- lapply(srv_data_curves, function(x) {
      knots[-1] - 0.5 * diff(knots)
    })
  }
  else {
    q_m <- lapply(1:length(srv_data_curves), function(j) {
      old_diff <- diff(c(srv_data_curves[[j]]$t, 1))
      new_diff <- diff(t_optims[[j]])
      as.matrix(srv_data_curves[[j]][, -1]) * sqrt(old_diff/new_diff)
    })
    m <- lapply(t_optims, function(t_optim) {
      t_optim[-1] - 0.5 * diff(t_optim)
    })
  }

  #build datacurve ids
  ids <- lapply(1:length(m), function(j){
    id <- rep(j, length(m[[j]]))
  })
  #convert in long format
  ids_long <- do.call(c, ids)
  m_long <- do.call(c, m)
  q_m_long <- do.call(rbind, q_m)

  # Return
  q_m_long <- complex(real = q_m_long[,1], imaginary = q_m_long[,2])
  data.frame("id" = ids_long, "m_long" = m_long, "q_m_long" = q_m_long)
}


# Smooth cov surface using mgcv::bam with (skew)-symmetric tensor product p-splines and REML.
smooth_cov_surface <- function(cov_dat, knots, type, penalty, var_type, cluster){

  # Parameters for covariance smoothing
  cov.m = ifelse(type == "smooth", 0, -1)  # spline degree - 1
  cov.d = penalty # penalty
  cov.fx = ifelse(penalty == -1, TRUE, FALSE)
  cov.knots = get_knots(knots, type)
  cov.k = length(cov.knots) - cov.m - 2  # basis dimension

  # fit covariance surface and measurement-error variance.
  cov_fit <- list()
  if(var_type == "constant"){
    cov_fit$re <- mgcv::bam(
      Re(qq) ~ s(s, t, bs="symm", k = cov.k, m = c(cov.m, cov.d), fx = cov.fx, xt = list(skew = FALSE)) + st,
      data = cov_dat, method = "REML", knots=list(s = cov.knots, t = cov.knots, cluster = cluster)
    )
  } else if(var_type == "smooth") {
    cov_fit$re <- mgcv::bam(
      Re(qq) ~ s(s, t, bs="symm", k = cov.k, m = c(cov.m, cov.d), fx = cov.fx, xt = list(skew = FALSE))
      + s(t, by = st, bs = "ps", k = cov.k, m = c(cov.m, 1), fx = cov.fx),
      data = cov_dat, method = "REML", knots=list(s = cov.knots, t = cov.knots, cluster = cluster)
    )
  } else {
    cov_fit$re <- mgcv::bam(
      Re(qq) ~ s(s, t, bs="symm", k = cov.k, m = c(cov.m, cov.d), fx = cov.fx, xt = list(skew = FALSE)),
      data = cov_dat, method = "REML", knots=list(s = cov.knots, t = cov.knots, cluster = cluster)
    )
  }
  cov_fit$im <- mgcv::bam(
    Im(qq) ~ -1 + s(s, t, bs="symm", k = cov.k, m = c(cov.m, cov.d), fx = cov.fx, xt = list(skew = TRUE)),
    data = cov_dat, method = "REML", knots=list(s = cov.knots, t = cov.knots, cluster = cluster)
  )
  cov_fit
}


# Get coefficient matrix for pca
get_complex_coef_matrix <- function(cov_fit){
  beta.mat.re <- get_coef_matrix_from_model(cov_fit$re)
  beta.mat.im <- get_coef_matrix_from_model(cov_fit$im)
  matrix(complex(real = as.vector(beta.mat.re), imaginary = as.vector(beta.mat.im)), ncol = ncol(beta.mat.re))
}

get_coef_matrix_from_model <- function(model){
  coefs <- get_unconstrained_basis_coefs(model)
  s_ <- model$smooth[[1]]
  beta <- s_$Z %*% coefs
  F <- s_$bs.dim
  matrix(beta, nrow=F, ncol=F)
}

# Calculate the unconstrained basis coefs from constrained mgcv smooth.
get_unconstrained_basis_coefs <- function(model){
  s_ <- model$smooth[[1]]

  # Create toy data that ensures identifiability of trafo mat.
  knots <- s_$knots
  knots_mid <- knots[-length(knots)] + 0.5 * diff(knots)
  knots_mid <- knots_mid[-length(knots_mid)]
  knots_all <- unique(sort(c(knots,knots_mid)))
  toy_dat <- expand.grid(s = knots_all, t = knots_all)

  # Create toy design matrix and calculate trafo matrix.
  X <- mgcv::Predict.matrix(s_, toy_dat)
  X_ <- mgcv::PredictMat(s_, toy_dat)
  X_ <- if(s_$xt$skew == FALSE) {cbind(1, X_)} else {X_}  # ToDo: Check this!
  D <- solve(crossprod(X), crossprod(X, X_))

  # Apply trafo to model coefs.
  coef_idxs <- unique(c(1,s_$first.para:s_$last.para))  # Only intercept and tensor product smooth.
  D %*% stats::coef(model)[coef_idxs]
}

